eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+        message = self.create_signing_message(
+            block.slot_number,
+            parent_hashes,
+            shard_id,
+            shard_block_hash,
+            justified_slot,
+        )
+        sigs = [
+            bls.sign(
+                message,
+                private_key,
+            )
+        ]
+        aggregate_sig = bls.aggregate_sigs(sigs)
+
+        return self.get_attestation_record_class()(
+            slot=block.slot_number,
+            shard_id=shard_id,
+            oblique_parent_hashes=(),
+            shard_block_hash=shard_block_hash,
+            attester_bitfield=attester_bitfield,
+            justified_slot=justified_slot,
+            justified_block_hash=justified_block_hash,
+            aggregate_sig=aggregate_sig,
+        )
+
+    @classmethod
+    def aggregate_attestation_record(cls,
+                                     crystallized_state: CrystallizedState,
+                                     active_state: ActiveState,
+                                     block: BaseBeaconBlock,
+                                     votes: Iterable[Tuple[int, bytes, int]],
+                                     proposer_attestation: 'AttestationRecord',
+                                     cycle_length: int) -> 'AttestationRecord':
+        """
+        Aggregate the votes.
+
+        TODO: Write tests
+        """
+        # Get signing message
+        parent_hashes = get_hashes_to_sign(
+            active_state.recent_block_hashes,
+            block,
+            cycle_length,
+        )
+        message = cls.create_signing_message(
+            block.slot_number,
+            parent_hashes,
+            proposer_attestation.shard_id,
+            block.shard_block_hash,
+            crystallized_state.last_justified_slot,
+        )
+        # Update the bitfield and append the signatures
+        bitfield = proposer_attestation.bitfield
+        sigs = []
+        for (committee_index, sig, public_key) in votes:
+            if bls.verify(message, public_key, sig):
+                bitfield = set_voted(bitfield, committee_index)
+                sigs.append(sig)
+
+        return proposer_attestation.copy(
+            bitfield=bitfield,
+            sigs=bls.aggregate_sigs(sigs),
+        )
+
+    #
+    #
+    # Validation
+    #
+    #
+
+    #
+    # Parent block proposer validation
+    #
+    @classmethod
+    def validate_parent_block_proposer(cls,
+                                       crystallized_state: CrystallizedState,
+                                       block: BaseBeaconBlock,
+                                       parent_block: BaseBeaconBlock,
+                                       cycle_length: int) -> None:
+        if block.slot_number == 0:
+            return
+
+        block_committees_info = get_block_committees_info(
+            parent_block,
+            crystallized_state,
+            cycle_length,
+        )
+
+        if len(block.attestations) == 0:
```

Comment:
Rather than pre-check, I think the more idiomatic way to do this is:

```python
try:
    attestation = block.attestations[0]
except IndexError:
    raise ValidationError(...)
```